code/crates/core-consensus/src/handle/decide.rs

```diff
@@ -1,25 +1,67 @@
 use crate::prelude::*;
 
 #[cfg_attr(not(feature = "metrics"), allow(unused_variables))]
-pub async fn decide<Ctx>(
+pub async fn try_decide<Ctx>(
     co: &Co<Ctx>,
     state: &mut State<Ctx>,
     metrics: &Metrics,
-    consensus_round: Round,
-    proposal: SignedProposal<Ctx>,
 ) -> Result<(), Error<Ctx>>
 where
     Ctx: Context,
 {
-    let height = proposal.height();
-    let proposal_round = proposal.round();
-    let value = proposal.value();
+    if !state.driver.step_is_commit() {
+        return Ok(());
+    }
+
+    let height = state.driver.height();
+    let consensus_round = state.driver.round();
+
+    let Some((proposal_round, decided_value)) = state.decided_value() else {
+        return Err(Error::DecisionNotFound(height, consensus_round));
+    };
 
-    // We only decide proposals for the current height
-    assert_eq!(height, state.driver.height());
+    let decided_id = decided_value.id();
+
+    // Look for an existing certificate
+    let (certificate, extensions) = state
+        .driver
+        .commit_certificate(proposal_round, decided_id.clone())
+        .cloned()
+        .map(|certificate| (certificate, VoteExtensions::default()))
+        .unwrap_or_else(|| {
+            // Restore the commits. Note that they will be removed from `state`
+            let mut commits = state.restore_precommits(height, proposal_round, &decided_value);
 
-    // Clean proposals and values
-    state.remove_full_proposals(height);
+            let extensions = extract_vote_extensions(&mut commits);
+
+            // TODO: Should we verify we have 2/3rd commits?
```

Comment:
We can use [`verify_certificate`](https://github.com/informalsystems/malachite/blob/3b291f737ed93276b80d5bb23a76548bc25636fb/code/crates/core-consensus/src/handle/signature.rs#L48) which will yield a `VerifyCertificate` effect that will end up checking signatures etc. in https://github.com/informalsystems/malachite/blob/c3e0beafc510faf0d326edefd17a02e675387267/code/crates/core-types/src/signing.rs#L224

Reply:
Added the certificate verification in ef7d5aac9b92f929243ead5547f3af6d066d6cee